Config experiment2

rjwats opened the pull request:
Another WIP: changes which could allow external configuration via configuration objects.

rjwats commented:
@justoke 

Second PR showing an extended approach which can use settings objects in addition to Strings and JSON objects.

Still not 100% sold on the changes here. The use of copy constructors might be bad from a memory perspective, without them though it's difficult to protect the settings classes from having their properties changed unbeknown to them though.

From an API perspective it's nice to be able to deal with configuration objects though:
`
// grab the config object from the service
WiFiSettings settings = _esp8266React->getWiFiSettingsService()->fetch();

// do something with the config
Serial.print("The ssid is:");
Serial.println(settings.ssid);

// set some new settings
settings.ssid="SomeNewSSID";
_esp8266React->getWiFiSettingsService()->update(settings);
`

Let me know your thoughts.

justoke commented:
I like this implementation. Memory impact aside, this gives access to the settings wherever they may be and addresses the need for the consumer - demoProject - to access settings across services. 

As access to the settings in each service is via the _esp8266React instance, this feels like it wraps it up nicely. Returning a strongly typed instance with public properties removes the need for dealing with the JSON documents - this is a lower level which people using the library would not have to delve into.

Always trade-offs with exposing a two way mechanism which is outside of the services, but seeing as the demoProject can be notified when settings are changed in a particular service, I'd say this would be the most common use case - consuming settings across services and dealing with changes if and when made via the UI. I'd probably create private properties for all the settings that I cared about in the demoProject constructor or add a setup method and only refresh if the settings were changed.

proddy commented:
I like it too and a much needed feature. I'm still not 100% onboard with the way settings are stored but thats a personal preference from years struggling with ESP and how they handle I/O. First off Arduino Strings are evil (see https://github.com/rjwats/esp8266-react/issues/79) so best practice is to avoid duplication specially when initiating JSON objects. Secondly I would have created a single JSON doc for the whole config and store as a messagepack. This prevents a lot of reading from many individual SPIFFS files and writing to them physically wears out the flash ram over time. SPIFFS operations also interfere with low level UART interrupts which is why I opted not to use them extensively in my iot projects. Additionally if you want a backup/restore function it's a lot easier when there's a single document to parse.

rjwats commented:
I was under the impression that the static wear levelling features of the SPIFFS implementation should significantly reduce the speed at which you will wear out the flash when writing using a truncating mode file handle (the settings service class uses "w"). I believe you have to be much more careful when opening files with "r+" and using seek() and write() as this bypasses the wear levelling. 

I also was worried about wear on the flash and looked into this before. I was looking at a way to [compress log data](https://github.com/rjwats/esp8266-react/blob/fermentation-chamber/src/fermentation-chamber/CircularLog.h) so I could store and retrieve historical temperature data from SPIFFS. I eventually decided that an implementation that wrote a single data point ("slot") at a time to the flash would be OK. Given the rate I was gathering data it would take ~27 years before the flash reached it's write limits - still working fine to this day!! (I should really update my fermentation chamber to the latest branch!)

I had an idea about the backup requirement too. I like that feature. A neat option would be for the service settings classes to register themselves with a backup service (internally holding hashmap of string -> pointer to SettingsService instance) which can handle backup + restore by delegating to the updateFromDocument and fetchAsDocument functions of the appropriate classes:

```
backupService.register("wifi", &wifiSettingsService);
backupService.register("ota", &otaSettingsService);
// etc...
```

This has the benefit of being re-usable outside the framework core and keeps things encapsulated. With that approach the "project" could register its own settings for backup/restore without making any framework changes.

proddy commented:
yup, you're right about the static wear levelling. On another project I had previously stored a rotating log in SPIFFS capturing all events and ran into problems - but was using seek() and fetching in chunks. Eventually, I just moved it all to syslog.

Subscribing to a backup service as you described is a great way too.

You're adding good stuff - do you need any help anywhere?

rjwats commented:
Thanks for the feedback on this!

Have added some documentation and tidy ups to this tonight - little bit more testing tomorrow then it's going in.

justoke commented:
Thank you - been burning the midnight oil I see, looking forward to these enhancements.

justoke commented:
> Hi, I had an issue with this which once. when I tried to build the framework after a new platform release. It seemed to be down to the declaration order of the headers though I'd love to be able to be more precise. This commit fixed the build before - I managed to fix it without explicitly adding FreeRTOS.h: [14f50c1](https://github.com/rjwats/esp8266-react/commit/14f50c1e31b079699323f05a7dc1623a3d2dd261) Maybe there's something in that commit which could help.
> […](#)
> On Thu, Jan 16, 2020 at 10:58 PM justoke ***@***.***> wrote: Resolved - had to add as follows to SettingsService.h #ifndef SettingsService_h #define SettingsService_h #include <FreeRTOS.h> ... — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <[c66968a?email_source=notifications&email_token=AAKE4VG2QRQLOW7KSWOLYTDQ6DRCVA5CNFSM4KH3SV5KYY3PNVWWK3TUL52HS4DFVVBW63LNNF2EG33NNVSW45FKMNXW23LFNZ2F62LEZYBDEGZQ#commitcomment-36838192](https://github.com/rjwats/esp8266-react/commit/c66968ae3ecb3b8d1ad8c94fd5db97770eff22e4?email_source=notifications&email_token=AAKE4VG2QRQLOW7KSWOLYTDQ6DRCVA5CNFSM4KH3SV5KYY3PNVWWK3TUL52HS4DFVVBW63LNNF2EG33NNVSW45FKMNXW23LFNZ2F62LEZYBDEGZQ#commitcomment-36838192)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAKE4VCJNFBQMRZ53HNK2TLQ6DRCVANCNFSM4KH3SV5A> .

I think this is the issue. It depends on where WiFi.h is declared, when AsyncTCP.h is used. I thought I saw you mention this somewhere but I've lost the thread. Anyway, I went through the header files and corrected these and removed FreeRTOS.h from all the places I added it and it worked.

This is good:
```
#ifdef ESP32
#include <WiFi.h>
#include <AsyncTCP.h>
#elif defined(ESP8266)
#include <ESP8266WiFi.h>
#include <ESPAsyncTCP.h>
#endif
```

And this is not so good:
```
#ifdef ESP32
#include <AsyncTCP.h>
#include <WiFi.h>
#elif defined(ESP8266)
#include <ESP8266WiFi.h>
#include <ESPAsyncTCP.h>
#endif
```

rjwats commented:
Yes, that's right.

The auto-formatter was re-arranging the order of the includes, and the
problem was coming back! :)

For future reference, this clang format setting disables re-ordering of
includes:

https://github.com/rjwats/esp8266-react/commit/0ca9530afa15546518dfb42ab10f6569deebff08




On Sat, Feb 1, 2020 at 1:49 PM justoke <notifications@github.com> wrote:

> Hi, I had an issue with this which once. when I tried to build the
> framework after a new platform release. It seemed to be down to the
> declaration order of the headers though I'd love to be able to be more
> precise. This commit fixed the build before - I managed to fix it without
> explicitly adding FreeRTOS.h: 14f50c1
> <https://github.com/rjwats/esp8266-react/commit/14f50c1e31b079699323f05a7dc1623a3d2dd261>
> Maybe there's something in that commit which could help.
> … <#m_7966798049743372920_>
> On Thu, Jan 16, 2020 at 10:58 PM justoke *@*.***> wrote: Resolved - had
> to add as follows to SettingsService.h #ifndef SettingsService_h #define
> SettingsService_h #include <FreeRTOS.h> ... — You are receiving this
> because you authored the thread. Reply to this email directly, view it on
> GitHub <
> c66968a?email_source=notifications&email_token=AAKE4VG2QRQLOW7KSWOLYTDQ6DRCVA5CNFSM4KH3SV5KYY3PNVWWK3TUL52HS4DFVVBW63LNNF2EG33NNVSW45FKMNXW23LFNZ2F62LEZYBDEGZQ#commitcomment-36838192
> <https://github.com/rjwats/esp8266-react/commit/c66968ae3ecb3b8d1ad8c94fd5db97770eff22e4?email_source=notifications&email_token=AAKE4VG2QRQLOW7KSWOLYTDQ6DRCVA5CNFSM4KH3SV5KYY3PNVWWK3TUL52HS4DFVVBW63LNNF2EG33NNVSW45FKMNXW23LFNZ2F62LEZYBDEGZQ#commitcomment-36838192>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AAKE4VCJNFBQMRZ53HNK2TLQ6DRCVANCNFSM4KH3SV5A
> .
>
> I think this is the issue. It depends on where WiFi.h is declared, when
> AsyncTCP.h is used. I thought I saw you mention this somewhere but I've
> lost the thread. Anyway, I went through the header files and corrected
> these and removed FreeRTOS.h from all the places I added it and it worked.
>
> This is good:
>
> #ifdef ESP32
>
> #include <WiFi.h>
>
> #include <AsyncTCP.h>
>
> #elif defined(ESP8266)
>
> #include <ESP8266WiFi.h>
>
> #include <ESPAsyncTCP.h>
>
> #endif
>
>
> And this is not so good:
>
> #ifdef ESP32
>
> #include <AsyncTCP.h>
>
> #include <WiFi.h>
>
> #elif defined(ESP8266)
>
> #include <ESP8266WiFi.h>
>
> #include <ESPAsyncTCP.h>
>
> #endif
>
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/rjwats/esp8266-react/pull/84?email_source=notifications&email_token=AAKE4VAGOHJSPUZ32OUT4C3RAV4VXA5CNFSM4KMJ6UUKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKQ5RBQ#issuecomment-581032070>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKE4VEEQU7X44ZIY4MDP5DRAV4VXANCNFSM4KMJ6UUA>
> .
>
